src/catchup/CatchupManagerImpl.cpp

```diff
@@ -119,26 +129,32 @@ CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
         logAndUpdateCatchupStatus(true);
     }
 
+    // Always skip old ledgers
     uint32_t lastReceivedLedgerSeq = ledgerData.getLedgerSeq();
+    if (lastReceivedLedgerSeq <= *mLastQueuedToApply)
+    {
+        // If LCL is already at-or-ahead of the ledger we just received from the
+        // network, we're up to date. Return early, nothing to do.
+        CLOG_INFO(
+            Ledger,
+            "Skipping close ledger: local state is {}, more recent than {}",
+            *mLastQueuedToApply, ledgerData.getLedgerSeq());
+        return true;
+    }
+
+    // Always add a newer ledger, maybe apply
+    mSyncingLedgers.emplace(lastReceivedLedgerSeq, ledgerData);
     mLargestLedgerSeqHeard =
         std::max(mLargestLedgerSeqHeard, lastReceivedLedgerSeq);
 
     // 1. CatchupWork is not running yet
-    // 2. CatchupManager received  ledger that was immediately applied by
+    // 2. CatchupManager received  ledger that should be immediately applied by
     // LedgerManager: check if we have any sequential ledgers.
     // If so, attempt to apply mSyncingLedgers and possibly get back in sync
-    if (!mCatchupWork && lastReceivedLedgerSeq ==
-                             mApp.getLedgerManager().getLastClosedLedgerNum())
+    if (!mCatchupWork && lastReceivedLedgerSeq == *mLastQueuedToApply + 1)
```

Comment:
Yeah, so the condition we have in master is really confusing. It works though because of the weird separation of responsibilities between LM and CM. LedgerManager special cases LCL+1 case, and applies it immediately before handing the ledger to CM. So by the time CM gets the ledger, `lastReceivedLedgerSeq == mApp.getLedgerManager().getLastClosedLedgerNum()` condition is met, and CM can try to apply the remaining buffered ledgers. I believe the proposed change here is more intuitive: LM now always hands the ledger to CM, and CM can decide what to replay. This means that the condition here changes to `lastReceivedLedgerSeq == *mLastQueuedToApply + 1`, or in plain English "did we receive a ledger that can be applied sequentially"